FFmpeg Win64: Use optional shared DLL method and some general maintenance

Zexaron opened the pull request:
I thought it would be a good first step for future framedumping improvements.

- Switch to a different method of using FFmpeg, requires runtime DLLs, no custom-build integration.
- DelayedLoading ensures DLLs are optional, only required when dumping frames.
- Updated to FFmpeg 4.0 (includes, etc)
- Rename `Externals/ffmpeg` folder to `Externals/ffmpeg-win64` to make that more obvious
- Missing FFmpeg License and Readme files added


**Detailed explanation here:**
[Dolphin Forums - Framedumping Research](https://forums.dolphin-emu.org/Thread-researching-framedumping-issues-expectation-and-potential-solutions?pid=473027#pid473027)

degasus commented on Source/Core/DolphinQt2/MenuBar.cpp:
```diff
@@ -706,11 +729,28 @@ void MenuBar::AddMovieMenu()
+#if defined(_WIN32)
+  if (FFmpeg_Win32_Exists() == true)
+  {
+    auto* dump_frames = movie_menu->addAction(tr("Dump Frames"));
```
Can you write the macros in a way which doesn't need this code twice?

lioncash commented on Source/Core/Common/CommonPaths.h:
```diff
@@ -7,6 +7,7 @@
 // Directory separators, do we need this?
 #define DIR_SEP "/"
 #define DIR_SEP_CHR '/'
+#define DIR_SEP_WIN32 "\\"
```
I'm extremely skeptical this is even necessary.

Zexaron replied:
Well, I had 5 cases (5 files) and there's more characters that the two which already exist, I'll wait a bit for what others think about this.

JosJuice replied:
Please remove this.

lioncash commented on Source/Core/DolphinQt2/MenuBar.cpp:
```diff
@@ -53,6 +55,27 @@
 #include "UICommon/GameFile.h"
 
+#if defined(_WIN32)
+bool FFmpeg_Win32_Exists()
```
Underscores should be removed. This should also be a static function.

RisingFog commented:
Could you branch out the External changes into a different commit?

lioncash commented on Source/Core/DolphinQt2/MenuBar.cpp:
```diff
@@ -53,6 +55,27 @@
+  const std::string swresample = File::GetExeDirectory() + DIR_SEP_WIN32 + "swresample-3.dll";
+  const std::string swscale = File::GetExeDirectory() + DIR_SEP_WIN32 + "swscale-5.dll";
+
+  if (File::Exists(avcodec) && File::Exists(avformat) && File::Exists(avutil) &&
```
This can just be compressed into:

```cpp
return File::Exists(avcodec) && File::Exists(avformat) && File::Exists(avutil) &&
       File::Exists(swresample) && File::Exists(swscale);
```

lioncash commented on Source/Core/DolphinQt2/MenuBar.cpp:
```diff
@@ -706,11 +729,28 @@ void MenuBar::AddMovieMenu()
   movie_menu->addSeparator();
 
+#if defined(_WIN32)
+  if (FFmpeg_Win32_Exists() == true)
```
`== true` is unnecessary

lioncash commented on Source/Core/DolphinQt2/MenuBar.cpp:
```diff
@@ -924,6 +964,38 @@ void MenuBar::UpdateToolsMenu(bool emulation_started)
 
+#if defined(_WIN32)
+#include <Common/FileUtil.h>
+void MenuBar::DumpFrames_FFmpegNotFound_Win32()
```
Underscores should be removed.

lioncash commented on Source/Core/DolphinQt2/MenuBar.cpp:
```diff
@@ -924,6 +964,38 @@ void MenuBar::UpdateToolsMenu(bool emulation_started)
+#include <Common/FileUtil.h>
+void MenuBar::DumpFrames_FFmpegNotFound_Win32()
+{
+  const QString dol_root_dir = tr(File::GetExeDirectory().c_str());
```
Why is `tr` used here? This isn't translateable in any reasonable means. It's a directory path.

JosJuice replied:
You probably want `QString::fromStdString` here.

Zexaron replied:
This one took me some time to figure out, persistance paid off and after lots of combinations this is what worked. If I remove the tr() it it will not compile: `Error C2248 - 'QString::QString': cannot access private member declared in class 'QString'`
@JosJuice I'll try that.

RisingFog commented:
I'm a +1 on this, and it has always been something that I wanted to have done with ffmpeg on Dolphin.

This should cut out not only custom building the dll files, but also distributing them and opens up the possibility of allowing any supported codec being used for frame dumping.

Zexaron commented:
@RisingFog I just squashed them before uploading this branch, I had them separate, I've read some stuff and I guess it confused me and I squashed too early. I still have the old branches and backup copies, but you're suggesting I can just split this here into two commits? I didn't know that's possible so I'll go read up on that.

That's said, I have a bit of a question regarding my rename to the `ffmpeg-win64` - the `include` files (or how they're properly called) as in relation to platform specifics, are those really only used for Windows? But they do come from the same windows-specific package as the .lib and .dll

RisingFog commented:
@Zexaron They're only required for Windows, Linux/macOS grabs the dependencies on compile from the OS directly (if the dependencies are installed/available).

Zexaron commented:
@RisingFog Just to clear out, If I deleted the `Externals\ffmpeg\include` folder, would the solution compile successfully for Linux/macOS on the buildbot here, for example?

@RisingFog I've read up on commit splitting, because there is some hazards and I don't want to mess this PR up. It may take some time because I should do a test to another branch. I'm aware that more commits is easier for reviewing, but in this case does it also matter when merging. I might then just keep amending changes (not many of them) and do a commit split later once I figure it out better, right?

BhaaLseN commented:
Splitting is fairly easy; open git gui, check "amend last commit" and remove everything that isn't the folder rename - then commit it with a new message. Everything else thats left is the other changes.

Zexaron commented:
@BhaaLseN Thanks for the tip. Unfortunately I've unticked the GUI component when installing the 2.17 version a few months ago. I already plan to include it when I do a reinstall, but probably not a good idea to do that in the middle of this PR, IMO.

RisingFog commented:
@Zexaron It would compile on linux/macOS without that folder.

BhaaLseN commented:
I wasn't aware you could actually _not_ install `git gui`. Do you mean you're doing this all from cli with no gui whatsoever? In that case: `git reset HEAD~1` and re-`git add` the files related to the renaming.

Zexaron commented:
Fixed all of these points but commit split would have to wait until tomorrow at earliest, as it would probably stretch deep into the night.

@BhaaLseN It's a bit of a long story, a bit offtopic but in short, in the beginning last year I thought GitForWindows was it's own thing (some unreliable random source) but then official git-scm links to it, but I thought there were two for windows. I used it a bit before, but I settled with Bash, and after a fresh Win10 install 3 months ago I remember unticking it in the installer, since then I wanted to try the GUI again, but I couldn't find it in Program Files. However I forgot to confirm with a search, which I did right now, and I found git-gui.exe, why on earth is it not in the root directory like git bash and cmd. So maybe there's 2 different GUIs after all, in one program. No need to worry, I'll figure it all out tomorrow, too late today.

shuffle2 commented:
IMO it would be nicer to just have a yes/no click-through which would download and extract the needed files (maybe drop back to the instructions given in the current state if it fails). But that can be done in a different PR (if anyone wants to implement it).

Zexaron commented:
@shuffle2 That was the original plan indeed. The choice is to use this as is, temporairly, or wait until that is done separately and then merge both at a later date?

RisingFog commented on Source/Core/DolphinQt2/MenuBar.cpp:
```diff
@@ -706,11 +722,28 @@ void MenuBar::AddMovieMenu()
-          [](bool value) { SConfig::GetInstance().m_DumpFrames = value; });
+  static bool ffmpeg_notfound_win32 = false;
+#if defined(_WIN32)
+  if (FFmpegWin32Exists() == false)
```
This can be changed to if (!FFmpegWin32Exists())

RisingFog commented on Source/Core/DolphinQt2/DolphinQt2.vcxproj:
```diff
@@ -40,8 +40,9 @@
   <ItemDefinitionGroup>
     <Link>
-      <AdditionalLibraryDirectories>$(ExternalsDir)ffmpeg\lib;%(AdditionalLibraryDirectories)</AdditionalLibraryDirectories>
+      <AdditionalLibraryDirectories>$(ExternalsDir)ffmpeg-win64\lib;%(AdditionalLibraryDirectories)</AdditionalLibraryDirectories>
```
I'm almost certain that we don't need "$(ExternalsDir)ffmpeg-win64\lib;" anymore now that we're dynamically linking.

RisingFog commented on Source/Core/DolphinQt2/DolphinQt2.vcxproj:
```diff
@@ -40,8 +40,9 @@
     <Link>
-      <AdditionalLibraryDirectories>$(ExternalsDir)ffmpeg\lib;%(AdditionalLibraryDirectories)</AdditionalLibraryDirectories>
+      <AdditionalLibraryDirectories>$(ExternalsDir)ffmpeg-win64\lib;%(AdditionalLibraryDirectories)</AdditionalLibraryDirectories>
       <AdditionalDependencies>avrt.lib;iphlpapi.lib;winmm.lib;setupapi.lib;opengl32.lib;glu32.lib;rpcrt4.lib;comctl32.lib;avcodec.lib;avformat.lib;avutil.lib;swresample.lib;swscale.lib;Shlwapi.lib;%(AdditionalDependencies)</AdditionalDependencies>
```
avcodec.lib, avformat.lib, avutil.lib, swreample.lib, and swsacale.lib can also be removed from here.

RisingFog commented on Source/Core/DolphinWX/DolphinWX.vcxproj:
```diff
@@ -38,8 +38,9 @@
   <ItemDefinitionGroup>
     <Link>
-      <AdditionalLibraryDirectories>$(ExternalsDir)ffmpeg\lib;%(AdditionalLibraryDirectories)</AdditionalLibraryDirectories>
+      <AdditionalLibraryDirectories>$(ExternalsDir)ffmpeg-win64\lib;%(AdditionalLibraryDirectories)</AdditionalLibraryDirectories>
```
Here as well.

RisingFog commented on Source/Core/DolphinWX/DolphinWX.vcxproj:
```diff
@@ -38,8 +38,9 @@
     <Link>
-      <AdditionalLibraryDirectories>$(ExternalsDir)ffmpeg\lib;%(AdditionalLibraryDirectories)</AdditionalLibraryDirectories>
+      <AdditionalLibraryDirectories>$(ExternalsDir)ffmpeg-win64\lib;%(AdditionalLibraryDirectories)</AdditionalLibraryDirectories>
       <AdditionalDependencies>avrt.lib;iphlpapi.lib;winmm.lib;setupapi.lib;opengl32.lib;glu32.lib;rpcrt4.lib;comctl32.lib;avcodec.lib;avformat.lib;avutil.lib;swresample.lib;swscale.lib;Shlwapi.lib;%(AdditionalDependencies)</AdditionalDependencies>
```
Here too.

RisingFog commented on Source/UnitTests/UnitTests.vcxproj:
```diff
@@ -45,8 +45,9 @@
         dolphin codebase.
         -->
-      <AdditionalLibraryDirectories>$(ExternalsDir)ffmpeg\lib;%(AdditionalLibraryDirectories)</AdditionalLibraryDirectories>
+      <AdditionalLibraryDirectories>$(ExternalsDir)ffmpeg-win64\lib;%(AdditionalLibraryDirectories)</AdditionalLibraryDirectories>
```
And here.

RisingFog commented on Source/UnitTests/UnitTests.vcxproj:
```diff
@@ -45,8 +45,9 @@
         -->
-      <AdditionalLibraryDirectories>$(ExternalsDir)ffmpeg\lib;%(AdditionalLibraryDirectories)</AdditionalLibraryDirectories>
+      <AdditionalLibraryDirectories>$(ExternalsDir)ffmpeg-win64\lib;%(AdditionalLibraryDirectories)</AdditionalLibraryDirectories>
       <AdditionalDependencies>avrt.lib;iphlpapi.lib;winmm.lib;setupapi.lib;opengl32.lib;glu32.lib;rpcrt4.lib;comctl32.lib;avcodec.lib;avformat.lib;avutil.lib;swresample.lib;swscale.lib;Shlwapi.lib;%(AdditionalDependencies)</AdditionalDependencies>
```
And here.

lioncash commented on Source/Core/DolphinQt2/MenuBar.cpp:
```diff
@@ -924,6 +957,38 @@ void MenuBar::UpdateToolsMenu(bool emulation_started)
 }
 
+#if defined(_WIN32)
+#include <Common/FileUtil.h>
```
This include isn't necessary, since it's already included at the top of the file.

lioncash commented on Source/Core/DolphinQt2/MenuBar.cpp:
```diff
@@ -53,6 +55,20 @@
+#if defined(_WIN32)
+static bool FFmpegWin32Exists()
+{
+  const std::string avcodec = File::GetExeDirectory() + "\\" + "avcodec-58.dll";
```
This can just use `DIR_SEP` or `DIR_SEP_CHR` from CommonPaths, it doesn't need to be a backslash.

Zexaron commented:
@RisingFog Those .lib dependencies are still required due to the way I implemented delayed loading support, much more code changes would have been required and a manual implementation of the LoadLibrary/GetProcAddress. Simply putting DLLs user /DELAYLOAD uses the helper function which does this under the hood, it's hidden, you won't see it in the source.

RisingFog commented:
@Zexaron Alright, I didn't see that the new lib files were added in the diff. Ignore my review then for that.

Zexaron commented:
@RisingFog Oh ok, maybe if tick the "Approve Changes", might get rid of them.

JosJuice commented:
@Zexaron No, reviews don't work that way.

JosJuice commented on Source/Core/DolphinQt2/MenuBar.cpp:
```diff
@@ -924,6 +957,37 @@ void MenuBar::UpdateToolsMenu(bool emulation_started)
+  {
+    QClipboard* clipboard = QApplication::clipboard();
+    QString ffmpeg_win64_shared_url =
+        tr("https://ffmpeg.zeranoe.com/builds/win64/shared/ffmpeg-4.0-win64-shared.zip");
```
Don't use `tr` here.

JosJuice commented on Source/Core/DolphinQt2/MenuBar.h:
```diff
@@ -203,6 +207,9 @@ class MenuBar final : public QMenuBar
   QAction* m_recording_stop;
   QAction* m_recording_read_only;
+#ifdef _WIN32
+  QAction* m_dump_frames_noff_win32;
```
`noff` isn't an understandable word.

Zexaron commented:
@shuffle2 AFAIK delayloading just uses the helper function which does the LoadLibrary/GetProcAddress automatically, but the point is to have it delay loaded, not startup loaded, in other words, optional, if it's that easy just adding some DLL filenames to a setting it's well worth it.

Zexaron commented:
I wanted to rebase, just had a BSOD while in interactive rebase, why is my Win10 doing this to me now. Index file got corrupt for the whole Dolphin repo, sorry this is going to take some time and I'm busy with other health stuff. I might have to first fix the BSODs I had a few, it could be HW failing too.

There was no action happening, it was just in the middle of the rebase waiting for manual edits, index shuldn't be that easily corruptible, this is some serious issue if just an untidy shutdown of bash/cmd/mintty would cause this. The index file shows NULNULNUL ... and same thing happened on an untidy shutdown before without any BSODs.

Techjar commented:
I've had that happen before as well, my guess is it has something to do with disk write caching.

spycrab commented on Source/Core/DolphinQt2/AboutDialog.cpp:
```diff
@@ -39,6 +39,9 @@ AboutDialog::AboutDialog(QWidget* parent) : QDialog(parent)
   // can be downloaded at no cost, but that's not what this message says.)
   text.append(medium + tr("Dolphin is a free and open-source GameCube and Wii emulator.") +
               QStringLiteral("</p>"));
+  text.append(medium +
```
Did we do that for any other library we use that has an LGPL license?

Zexaron replied:
Yeah, I did thought about that later when some folks explained how it's not an issue if both programs are using the GPL licensing family. But it woud look and feel cleaner if maybe there would be some list, just for user info if not for legal, a button with an additional dialog could be one option which would list all the libraries and could add legal/homepage info/url as an added bonus.

spycrab replied:
This has nothing to do with this PR though and should be discussed separately.

Zexaron commented:
Looks like thing are going a bit better with my PC stability, I've imaged my Win10 off the Samsung 860 EVO 250GB to a similarly fresh 850 EVO 250GB which I had laying around meant to put a fresh install of Win7 on, and put it on another sata port as well as different sata cable, been working for 2 days now, but not so fast, still have to do more checks/test and just giving it some time, including using linux on a third 840 Pro SSD and putting that on the first port and the same cable, if linux has a kernel panic then it has to be the port or the sata cable or something in the controller/motherboard, ruling out the 860 as the culprit.

I know it's not directly relevant to this PR, but what I'm trying to get to is the schedule, if this PR won't get merged temporairly before the auto-ffmpeg-downloader PR is done (any volunteers?), then there's no point for me to rush here.

RisingFog commented:
Is there any update on this PR?

Zexaron commented:
@RisingFog I'm still around, althought unfortunately still dealing with the PC hardware troubleshooting and maintenance, various side-chores have opened up, had to be done first, other PCs got problems, I've had 5 hardware failures in less than 6 months, 3 warranties, after no issues for like 4 years, it's just bad coincidence, and I simply can't get to fixing the dolphin git repo until I get my PC back - I'm closer, I think one of the newer SSDs is randomly locking up after less than 2 months of use, since it's happeing on other SATA ports as well, now I have to test it on another PC, progress is slow becuase I have a lot of life stuff going on with the home improvement stuff, new appliances etc.

It's still up in the air for this PR, if there's an update, it would come in another PR, the updater/downloader for ffmpeg. But ofcourse I will resolve the conflicts in this PR ASAP.
I wouldn't want to mess with this one adding the updater here.

Plus, I was also going to update git and maybe learn/familiarize with one of the GUIs, so there's all of that as well.

RisingFog commented:
@Zexaron I could fork the PR and resolve the merge conflicts if you expect the issues to take a prolonged period of time to resolve on your end.

Zexaron commented:
@RisingFog With all the stuff I have going on, it would seem like a good idea if someone else took the updater task, I could probably get the issues resolved in a week or even earlier, I mean, the PC works fine on the other SSD I cloned to, typing from it right now, where I keep all my git stuff, on the main system drive yes, so I don't need anything else, but I tend to want to finish one big chore before I move to another.

In the meantime, would be helpful if you, would you want the updater to be part of dolphin code or an external app would get used? Then I've heard there's issues with admin-rights that have to be handled to do the auto-unzipping and file, so if it works here (where I use a heavily tweaked Win10 to get rid all of the nags, admin rights, among 100 other things) it may not work for everyone out there and it won't get merged, but not sure beyond that.

Doing it with an external utility might just need a few CMD scripts, since this is Windows only, command prompt is fine. But ofcourse it's more quality to have it done internally, it may just take more time for me to figure that out.

Zexaron commented:
@RisingFog There, resolved.

RisingFog commented:
@dolphin-emu-bot rebuild

Zexaron commented:
Allright, I didn't mean I'll be completely gone again, I did look a bit into some of the updater draft code I was doing back then and looking again into the curl and HttpRequest that's implemented already. I forgot exactly what barrier I hit that stopped me back then. Either someone takes it or I'll go slow and steady with it and do something eventually.

I'm not sure yet on those non-win build issue yet, but it has obviously something to do because code suppose to be WIN32 only.

Zexaron commented:
The FFmpeg license stuff was thankfully in it's own commit, so I figured out to remove it right here. I made changes to the WIN32 definitions, try now if it compiles on non-win32, I think it should.

Zexaron commented:
New: Managed to also include the downloading now, I figured it should be easier than what I first had in mind, but this is all inside the `MenuBar.cpp`.
It's not completely polished, no unzip yet, it just throws the zip into root install dir, the in-progress message doesn't show correctly and some other stuff, I've included it earlier not have to checkout constantly back and forth while I keep fixing the conflicts.

However I was able to fix the LINK errors here, now let's see if the other platforms work.

Zexaron commented:
Looks like some zip external would be needed, zlib can't handle .zip file format. Which one do you guys suggest ?

delroth commented:
Re: auto-downloading, I would rather we keep that for later. Instead, let's make the UI more user friendly -- having a link to documentation about how to install the required runtime libs in the error dialog.

There is a lot of complexity for auto-downloading (as mentioned above: permission problems, backwards and forwards compatibility with whatever we use as the serving backend for the files, etc.) and I'm not sure the small amount of users planning to use this (as opposed to just pointing OBS at the Dolphin window) warrants it.

Can you work with @MayImilae and @JMC47 to have a wiki page on video/audio dumping with Dolphin, which details the install instructions for the runtime libs and also mentions simpler alternatives (as mentioned, simply pointing OBS at Dolphin)?

Zexaron commented:
So moving the bulk of the instructions to the wiki ... didn't thought of that, I think it make sense yes because of the size and it could be better explained and managed there. Now I remember I spent a bit too much time just formatting the thing.

However, on the forums I saw the talk about framedumping about offline quality (lossless even), rather than streaming, those that are streaming (sacrificing quality in many ways) probably are using OBS already anyway, so this shouldn't be that down the totem pole.

I have reverted to the warning, replacing it with the wiki page that ofcourse isn't the final URL but I hope I hit the nail straight on. For now I left the clipboard URL copying in while I figure out the clickable hyperlink, I took a look at it a while ago and I think it's not that straightforward, especially because Qt docs and other help, AFAIK, expect an environment with QtCreator and XML. Unless I missed to find an example already implemented in Dolphin?

delroth commented:
You can put actual clickable links in a Qt dialog box (enabling rich text + using HTML tags).

Zexaron commented:
Allright, I figured it out.

Currently I left in the full URL print and the clipboard for backup incase there some kind of issue with hyperlink clickaibility or browser/OS incompatibility.

Clipboard shouldn't be a nag because it's a custom button now, doesn't need a text question.

The backup actually came nicely, FWIW, because the link location cuts off the URL parameter after = sign, so it redirects to the main page instead of that subpage, I couldn't figure this last bit out, pushing now, dealing with that later.

(Hmm now I'm wondering if I could make the dialog to close when hyperlink is clicked)

![ffmpegupd_shareddll_gui_v3](https://user-images.githubusercontent.com/20824154/46588584-225e0d00-ca9f-11e8-8f5e-579e3760631f.png)

delroth commented:
You're not quoting the url in your <a> tag. Should be:

    <a href="...">foo</a>

mbc07 commented:
Ok, wiki maintainer here, that will sound like nitpicking but here we go: while I'm OK in having the bulk of instructions in how to download and "install" the DLLs as a wiki page that the Dolphin application links to (especially when thinking in long term support), the chosen URL does **not** look good to me. 

Ideally, I think the dialog should just point to a generic URL (e.g. https://wiki.dolphin-emu.org/index.php?title=FFmpeg or maybe using dolp.in domain) that we then take care with an automatic redirection to the proper place, kinda like we currently do when users chose the "Wiki" option when right-clicking a game on the game list (it points to the game ID and we take care of that with a redirect), as it would give us wiki maintainer the most flexibility should we even need to move pages around without worrying breaking redirection in old builds that pointed to the old URL.

If this sounds a bit overkill (especially the dolp.in route, as it requires server-side changes), my vote is for at least using https://wiki.dolphin-emu.org/index.php?title=FFmpeg (and we then take care of that with an auto redirect on the wiki side), but if a hard link is absolutely necessary (I think it's not), then you should use https://wiki.dolphin-emu.org/index.php?title=Frame_Dumping, it fits better with the title of other tutorials/guides we already host at the wiki and it's also the proper term (framedumping, all together, seems wrong)...

Zexaron commented:
I mixed it up a bit, even tho I read it earlier but then forgot, now I thought wait, FFmpeg just happens to be the method frames get put together, not the feature it self, changing it to `..?title=Frame_Dumping` ofcourse. An updated screenshot shouldn't be necessary in this case.

mbc07 commented:
I requested the previous FFmpeg URL specifically to give us the best flexibility when dealing with the redirection on the wiki side and your previous commit addressed that perfectly. Please revert back to `...?title=FFmpeg`

And while you're at it, I think you don't need to expose the URL in the dialog as its text can't be selected and it already have a "Copy to clipboard" button. My suggestion:

```html
<p>One or more FFmpeg libraries is either missing, corrupt, inaccessible or invalid.
<br><br>Detailed frame dumping instructions are available 
<a href="https://wiki.dolphin-emu.org/index.php?title=FFmpeg">online on our wiki</a>.
```

Zexaron commented:
It wouldn't have made sense to me, but now I realize the error message is about FFmpeg it self ofcourse, I guess it leading to a FFmpeg-specific looking link makes sense in another way.

stenzek commented:
I'm also not convinced on the whole delay loading thing, as it's Windows-specific, and seems like a bit of a hack. What if frame dumping is enabled via the config, rather than the GUI? Does the program just crash if a symbol from a delayed loaded DLL can't be resolved?

I would rather go with dlopen/dlsym/LoadLibrary/GetProcAddress, that way we can make ffmpeg a runtime dependency across all platforms. I was discussing in another thread whether we should have a common "library loading helper" to abstract this away (we also need it for libvulkan). I don't think we're using _that_ many libavcodec exports which would have to be resolved..

delroth commented:
I can't think of a situation where you'd like ffmpeg delay-loading elsewhere than on Windows. On Linux it's not like people randomly go and install shared libraries but don't expect to have to recompile to use them. And there's the whole mess of ABI compatibility if you start wanting to make something generic... which I don't think is worth trying to solve.

stenzek commented:
This is the situation I was thinking of - make libavcodec an optional dependency, which could be added later without recompiling. But indeed, header versions/ABI differences could make it an issue.

That said, I'm still not sold on the linker-based method, due to the points above, but I'm not gonna bikeshed too much about it :P

Zexaron commented:
And I thought being windows-specific was a good thing, it made it a whole lot simpler imo.
The reasons for delayed-loading is for either or both of the two main issues I saw pretty early on, avoiding shipping ffmpeg with every release of dolphin and avoiding forcing all users to have it installed even if they're not going to use any functionality that requires FFmpeg.

Now if everyone doesn't care about the shipping size or the hosting space, there would be no need for delayed-loading nor user installation ... otherwise this PR would relieve around 1.5 MB.

Zexaron commented:
The wiki page has been set any it's under construction, in the meantime I added a temporary installation text from the original warning message. 

https://wiki.dolphin-emu.org/index.php?title=FFmpeg

Zexaron commented:
Warning for localization: trailing spaces here on purpose for a quick button width fix so that the border doesn't touch the text, otherwise I think a lot more code would be needed to adjust the width, unless if there's a shorter way that someone knows how to adjust the width without having to define the whole dialog element-by-element. 

https://github.com/dolphin-emu/dolphin/pull/7110/commits/9f2adb2f870cc70fdf1b426d79ee1e9c48c23b66#diff-8fa06c6cfb14789472b0c27be3da88b0R986

JosJuice commented:
If you want to warn localizers, make an i18n comment in the code.

Zexaron commented:
Done, althought it's a bit awkward positioning, would the `^^` trick work out ?

JosJuice commented:
The i18n comment has to be right above the string, not below. And using arrows won't make sense to localizers.

You should also make sure the comment refers to both the leading and trailing spaces, not just the trailing one.

Zexaron commented:
Corrections applied.

JosJuice commented:
No, the i18n comment has to be on the line right above the string, not 7 lines before.

Rukario commented:
Would it be alright if the "Copy to URL to clipboard" button is to be changed into just "Open Wiki" and it will open the URL in user-defined default browser, much like the "Wiki" button you see in context menu from game list? If you like this idea, may I also suggest drop the blue text, the button got it covered.

Zexaron commented:
I wasn't familiar this was this somewhere already in ... or I was but not from a developing viewpoint and just didn't realize. It probably has something to do with the OS's default browser, no? What if some users don't have that set up right, or got broken for some reason?

Some redundancy should be kept, no?

Zexaron commented:
I'd hope to have at least one redundancy, but whatever. Install instruction is a bit more critical if the user can't get to them IMO. We'll see what people think, or they may just like it how it is, but as far as me goes I'm fine with this as well.

The i18n line should be fixed now, I thought that wouldn't build. The wiki's also beginning to be shaped, but the primary thing, the install instructions, are more or less good for practical use as of now.

Zexaron commented:
@Rukario There that should make _everyone_ happy now !!!

![ffmpegupd_shareddll_gui_v6](https://user-images.githubusercontent.com/20824154/47171678-9acb9600-d309-11e8-8391-00bde57761e4.png)

spycrab commented on Source/Core/DolphinQt/MenuBar.cpp:
```diff
@@ -52,6 +53,20 @@
 #include "UICommon/GameFile.h"
 
+#if defined(_WIN32)
+static bool FFmpegWin32Exists()
```
This should probably go in ``UICommon`` somewhere

Zexaron replied:
Would this require including `UICommon.h` in the `MenuBar.cpp` if I put it in UICommon.cpp for example? You'd be okay with that?

Zexaron commented:
Okay, I put it in `UICommon.cpp` as it made most sense at first glance, which required including `UICommon.h`. Hopefully that's fine.

vadosnaprimer commented:
Won't [piping over to whatever version of ffmpeg is available on the system](https://github.com/clementgallet/libTAS/blob/master/src/library/encoding/AVEncoder.cpp) make it easier than linking a shared lib? That would allow to just use whatever commands are supported by the given ffmpeg build, nothing would have to be explicitly loaded, no internal dependencies to resolve (I hope I'm understanding this correctly).

JosJuice commented:
The Zexaron account doesn't exist anymore, meaning this PR can't be updated, so...

vadosnaprimer commented:
Could you still comment on my suggestion?

JosJuice commented:
It's not really the area of Dolphin I focus on, but... I guess it has the disadvantage of making things harder to set up for people who just want a quick and easy dump (not TASVideos.org quality) but doesn't really have disadvantages other than that? The disadvantage doesn't sound too bad to me considering that only a small subset of users use frame dumping, but I can't speak for what the other devs think.

vadosnaprimer commented:
If ffmpeg is not found, an error message with a link makes it rather trivial imo. On the other hand, having it in PATH permanently means you only download it once and all dolphin installations and portables can use it. It's also nice to have a couple default ffmpeg commands and a way to change them.

altimumdelta commented:
Intent to revive this on my own, anyone objecting?
